### DISCUSSION TITLE
nk.hilbert.HilbertIndex and nk.supervised.Supervised

### DISCUSSION THREAD

**Discussion by tortijazz** [netket/netket#902] [Q&A] [answered]
Hi everybody,

I'm having troubles reinterpreting an old script written in 2.0. It uses the nk.hilbert.HilbertIndex and  nk.supervised.Supervised functions, but these are no longer available in the new version, so my question is which functions can replace what the supervised and HilbertIndex functions did in the old Netket's version? If you could tell me what they were made for I would appreciate it.

```
import netket as nk 
from numpy import log

# 1D Lattice
g = nk.graph.Hypercube(length=10, n_dim=1, pbc=True)

# Hilbert space of spins on the graph
hi = nk.hilbert.Spin(s=0.5, graph=g)

# Ising spin Hamiltonian
ha = nk.operator.Ising(h=1.0, hilbert=hi)

# Perform Exact Diagonalization to get lowest eigenvector
res = nk.exact.lanczos_ed(ha, first_n=1, compute_eigenvectors=True)

# Store eigenvector as a list of training samples and targets
# The samples would be the Hilbert space configurations and
# the targets should be wavefunction amplitudes.
hind = nk.hilbert.HilbertIndex(hi)
h_size = hind.n_states
targets = [[log(res.eigenvectors[0][i])] for i in range(h_size)]
samples = [hind.number_to_state(i) for i in range(h_size)]

# Machine: Restricted Boltzmann machine
# with 20 hidden units
ma = nk.machine.RbmSpin(hilbert=hi, n_hidden=20)
ma. init_random_parameters(seed=1234, sigma=0.01)

# Optimizer
op = nk.optimizer.AdaMax()

# Supervised Learning module
spvsd=nk.supervised.Supervised(machine=ma, optimizer=op, batch_size=400, samples=samples, targets=targets)

# Run the optimization for 2000 iterations
spvsd.run(n_iter=2000, output_prefix=’test’, loss_function="Overlap_phi")
```

**Comment by gcarleo** [comment, marked as answer]
hi, concerning `HilbertIndex`, those methods have been just merged into the Hilbert class itself. 
You can see all the available methods just looking at our documentation, for example [here](https://www.netket.org/docs/_generated/hilbert/netket.hilbert.Spin.html) (you see we have things like states_to_numbers, numbers_to_states, all_states etc... )

Unfortunately supervised learning is not implemented in 3.0, but it should be relatively straightforward to do it yourself using the fact that all our models are JAX differentiable. At the moment we don't have plans to implement it, but I think that eventually it will appear also in 3.0

**Reply by tortijazz** [reply]
Thank you, Carleo. I appreciate your answer.
